File: core/src/main/java/com/gitlab/aecsocket/sokol/core/registry/Keyed.java
package com.gitlab.aecsocket.sokol.core.registry;

import net.kyori.adventure.text.Component;

import java.util.Locale;

public interface Keyed {

    String VALID = "abcdefghijklmnopqrstuvwxyz0123456789._-";

    static String validate(String key) throws ValidationException {
        for (int i = 0; i < key.length(); i++) {
            char c = key.charAt(i);
            if (!VALID.contains(""+c))
                throw new ValidationException(i, c);
        }
        return key;
    }

    String id();
    Component name(Locale locale);
}

File: paper/src/main/kotlin/com/gitlab/aecsocket/sokol/paper/util/Keyed.kt
package com.gitlab.aecsocket.sokol.paper.util

import com.gitlab.aecsocket.alexandria.core.keyed.Keyed
import net.kyori.adventure.key.InvalidKeyException
import net.kyori.adventure.key.Key
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.serialize.SerializationException
import java.lang.reflect.Type

fun validateStringKey(type: Type, node: ConfigurationNode): String {
    return try {
        Keyed.validate(node.key().toString())
    } catch (ex: Keyed.ValidationException) {
        throw SerializationException(node, type, "Invalid key", ex)
    }
}

fun validateNamespacedKey(type: Type, node: ConfigurationNode): Key {
    return try {
        Key.key(node.key().toString())
    } catch (ex: InvalidKeyException) {
        throw SerializationException(node, type, "Invalid namespaced key", ex)
    }
}
